src/compute-client/src/controller/instance.rs

```diff
@@ -2818,17 +2796,21 @@ impl<T: ComputeControllerTimestamp> ReplicaCollectionState<T> {
 
     /// Returns whether this collection is hydrated.
     fn hydrated(&self) -> bool {
-        self.hydrated
-    }
-
-    /// Marks the collection as hydrated and updates metrics and introspection accordingly.
-    fn set_hydrated(&mut self) {
-        if let Some(metrics) = &self.metrics {
-            let duration = self.created_at.elapsed().as_secs_f64();
-            metrics.initial_output_duration_seconds.set(duration);
-        }
-
-        self.hydrated = true;
+        // If the observed frontier is greater than the collection's as-of, the collection has
+        // produced some output and is therefore hydrated.
+        //
+        // We need to consider the edge case where the as-of is the empty frontier. Such an as-of
+        // is not useful for indexes, because they wouldn't be readable. For write-only
+        // collections, an empty as-of means that the collection has been fully written and no new
+        // dataflow needs to be created for it. Consequently, no hydration will happen either.
+        //
+        // Based on this, we could respond in two ways: * `false`, as in "the dataflow was never
+        // created" * `true`, as in "the dataflow completed immediately"
+        //
+        // Since hydration is often used as a measure of dataflow progress and we don't want to
+        // give the impression that certain dataflows are somehow stuck when they are not, we go
+        // with the second interpretation here.
+        self.as_of.is_empty() || PartialOrder::less_than(&self.as_of, &self.output_frontier)
```

Comment:
The comment's formatting somehow got lost: the `*` bullet points should start on new lines.

Reply:
Good catch! Pretty sure that happened because I carelessly used vim's `gq` command here 🙈